docs/proposals/pod-lifecycle-event-generator.md

```diff
@@ -0,0 +1,234 @@
+match the spec by performing a series of operations. It should be
+noted that this proposal does not intend to change this property --
+the sync pod routine would still perform all necessary checks,
+regardless of the event type. This trades some efficiency for
+reliablity and eliminate the need to build a state machine that is
+compatible with different runtimes.
+
+A pod worker's actions could lead to pod lifecycle events (e.g.,
+create/kill a container), which the worker would not observe until
+later. The pod worker should ignore such events to avoid unncessary
+work.
+
+For example, assume a pod has two containers, A and B. The worker
+
+ - Creates container A
+ - Receives an event `(ContainerStopped, B)`
+ - Recieves an event `(ContainerStarted, A)`
+
+
+The worker should ignore the `(ContainerStarted, A)` event since it is
+expected. Arguably, the worker could process `(ContainerStopped, B)`
+as soon as it receives the event, before observing the creation of
+A. However, it is desirable to wait until the expected event
+`(ContainerStarted, A)` is observed to keep a consistent per-pod view
+at the worker. Therefore, the control flow of a single pod worker
+should adhere to the following rules:
+
+1. Pod worker should process the events sequentially.
+2. Pod worker should not start syncing until it observes the outcome of its own
+   actions in the last sync to maintain a consistent view.
+
+In other words, a pod worker should record the expected events, and
+only wake up to perform the next sync until all expectations are met.
+
+ - Creates container A, records an expected event `(ContainerStarted, A)`
+ - Receives `(ContainerStopped, B)`; stores the event and goes back to sleep.
+ - Recieves `(ContainerStarted, A)`; clears the expectation. Proceeds to handle
+   `(ContainerStopped, B)`.
+
+In addition, whenever a pod worker encounters an error during the
+sync, it should set a timer to wake itself up after a short interval
+to sync again.  Similarly, we should set an expiration time for each
+expected events to prevent the worker from being stalled indefinitely
+by missing events.
+
+
+## Prototype PR
+
+I wrote a [scratch PR](https://issues.k8s.io/12829)
+implementing the basics of PLEG using cAdvisor's event stream for
+Docker, with a relist period of 30 seconds. It does not implement the
+control flow of pod worker yet, hence redundant syncs and sync errors
+are not handled properly.
+
+## TODOs
+
+Below is a list of issues to address. Details will be posted in
+separate issues soon.
+
+- SyncPods: Refactor and restructure the SyncPod routine to allow clear
+  expectations of actions [#12826](https://issues.k8s.io/12826).
+
+- Probing: Now that pod workers only wake up on events, we need to
+  rethink how liveness/readiness probing should be implemented. One
+  solution is to create a separate probing component that can be
+  plugged in to PLEG to generate corresponding pod lifecylce
```

Comment:
This is cool. I like the idea of a plugin to present the probe outcomes as events.